django/db/backends/sqlite3/_functions.py

```diff
@@ -0,0 +1,481 @@
+    create_deterministic_function('MD5', 1, _sqlite_md5)
+    create_deterministic_function('MOD', 2, _sqlite_mod)
+    create_deterministic_function('PI', 0, _sqlite_pi)
+    create_deterministic_function('POWER', 2, _sqlite_power)
+    create_deterministic_function('RADIANS', 1, _sqlite_radians)
+    create_deterministic_function('REPEAT', 2, _sqlite_repeat)
+    create_deterministic_function('REVERSE', 1, _sqlite_reverse)
+    create_deterministic_function('RPAD', 3, _sqlite_rpad)
+    create_deterministic_function('SHA1', 1, _sqlite_sha1)
+    create_deterministic_function('SHA224', 1, _sqlite_sha224)
+    create_deterministic_function('SHA256', 1, _sqlite_sha256)
+    create_deterministic_function('SHA384', 1, _sqlite_sha384)
+    create_deterministic_function('SHA512', 1, _sqlite_sha512)
+    create_deterministic_function('SIGN', 1, _sqlite_sign)
+    create_deterministic_function('SIN', 1, _sqlite_sin)
+    create_deterministic_function('SQRT', 1, _sqlite_sqrt)
+    create_deterministic_function('TAN', 1, _sqlite_tan)
+    # Don't use the built-in RANDOM() function because it returns a value
+    # in the range [-1 * 2^63, 2^63 - 1] instead of [0, 1).
+    connection.create_function('RAND', 0, random.random)
+    connection.create_aggregate('STDDEV_POP', 1, StddevPop)
+    connection.create_aggregate('STDDEV_SAMP', 1, StddevSamp)
+    connection.create_aggregate('VAR_POP', 1, VarPop)
+    connection.create_aggregate('VAR_SAMP', 1, VarSamp)
+
+
+def _sqlite_datetime_parse(dt, tzname=None, conn_tzname=None):
+    if dt is None:
+        return None
+    try:
+        dt = typecast_timestamp(dt)
+    except (TypeError, ValueError):
+        return None
+    if conn_tzname:
+        dt = dt.replace(tzinfo=timezone_constructor(conn_tzname))
+    if tzname is not None and tzname != conn_tzname:
+        tzname, sign, offset = split_tzname_delta(tzname)
+        if offset:
+            hours, minutes = offset.split(':')
+            offset_delta = timedelta(hours=int(hours), minutes=int(minutes))
+            dt += offset_delta if sign == '+' else -offset_delta
+        dt = timezone.localtime(dt, timezone_constructor(tzname))
+    return dt
+
+
+def _sqlite_date_trunc(lookup_type, dt, tzname, conn_tzname):
+    dt = _sqlite_datetime_parse(dt, tzname, conn_tzname)
+    if dt is None:
+        return None
+    if lookup_type == 'year':
+        return f'{dt.year:04d}-01-01'
+    elif lookup_type == 'quarter':
+        month_in_quarter = dt.month - (dt.month - 1) % 3
+        return f'{dt.year:04d}-{month_in_quarter:02d}-01'
+    elif lookup_type == 'month':
+        return f'{dt.year:04d}-{dt.month:02d}-01'
+    elif lookup_type == 'week':
+        dt = dt - timedelta(days=dt.weekday())
+        return f'{dt.year:04d}-{dt.month:02d}-{dt.day:02d}'
+    elif lookup_type == 'day':
+        return f'{dt.year:04d}-{dt.month:02d}-{dt.day:02d}'
+    raise NotImplementedError(f'Unsupported lookup type: {lookup_type!r}')
```

Comment:
> django.db.utils.OperationalError: user-defined function raised exception

This is true, but it's better than implicitly returning `NULL` and potentially bad results, no?

> we don't do this on other backends.

It's not applicable to other backends, because the DB server is the thing that will raise an error if such functions are called with invalid arguments, e.g. on postgres:

```
postgres=# select date_trunc('invalid', '2021-01-01'::date);
ERROR:  timestamp with time zone units "invalid" not recognized
```

Reply:
Error is raised only on PostreSQL. On Oracle and MySQL we simply return the same date. Even if we want to raise an exception on SQLite, `NotImplementedError` is probably not the best choice. It's not sth that will or may be implemented in the feature. I'd use `ValueError` as we do in similar cases.

Reply:
Sure, perhaps `ValueError` is better. I think PostgreSQL does the nice thing here - silently returning the value unchanged is ugly.

Given this is our own implementation, having a third way - returning `NULL` - isn't great. We should align to one of the other two behaviours, and raising an error seems best.

Reply:
Okay I've gone with `ValueError`, which is indeed more correct.